merkle-tree/hash-set/src/lib.rs

```diff
@@ -0,0 +1,1211 @@
+        assert_eq!(hs.contains(&element_2_6, 2).unwrap(), true);
+        assert_eq!(hs.contains(&element_2_8, 2).unwrap(), true);
+        assert_eq!(hs.contains(&element_2_9, 2).unwrap(), true);
+        hs.mark_with_sequence_number(&element_2_3, 2).unwrap();
+        hs.mark_with_sequence_number(&element_2_6, 2).unwrap();
+        hs.mark_with_sequence_number(&element_2_8, 2).unwrap();
+        hs.mark_with_sequence_number(&element_2_9, 2).unwrap();
+        assert!(matches!(
+            hs.insert(&element_2_3, 2),
+            Err(HashSetError::ElementAlreadyExists)
+        ));
+        assert!(matches!(
+            hs.insert(&element_2_6, 2),
+            Err(HashSetError::ElementAlreadyExists)
+        ));
+        assert!(matches!(
+            hs.insert(&element_2_8, 2),
+            Err(HashSetError::ElementAlreadyExists)
+        ));
+        assert!(matches!(
+            hs.insert(&element_2_9, 2),
+            Err(HashSetError::ElementAlreadyExists)
+        ));
+
+        let element_3_11 = 11.to_biguint().unwrap();
+        let element_3_13 = 13.to_biguint().unwrap();
+        let element_3_21 = 21.to_biguint().unwrap();
+        let element_3_29 = 29.to_biguint().unwrap();
+        hs.insert(&element_3_11, 2).unwrap();
+        hs.insert(&element_3_13, 2).unwrap();
+        hs.insert(&element_3_21, 2).unwrap();
+        hs.insert(&element_3_29, 2).unwrap();
+        assert_eq!(hs.contains(&element_3_11, 3).unwrap(), true);
+        assert_eq!(hs.contains(&element_3_13, 3).unwrap(), true);
+        assert_eq!(hs.contains(&element_3_21, 3).unwrap(), true);
+        assert_eq!(hs.contains(&element_3_29, 3).unwrap(), true);
+        hs.mark_with_sequence_number(&element_3_11, 3).unwrap();
+        hs.mark_with_sequence_number(&element_3_13, 3).unwrap();
+        hs.mark_with_sequence_number(&element_3_21, 3).unwrap();
+        hs.mark_with_sequence_number(&element_3_29, 3).unwrap();
+        assert!(matches!(
+            hs.insert(&element_3_11, 3),
+            Err(HashSetError::ElementAlreadyExists)
+        ));
+        assert!(matches!(
+            hs.insert(&element_3_13, 3),
+            Err(HashSetError::ElementAlreadyExists)
+        ));
+        assert!(matches!(
+            hs.insert(&element_3_21, 3),
+            Err(HashSetError::ElementAlreadyExists)
+        ));
+        assert!(matches!(
+            hs.insert(&element_3_29, 3),
+            Err(HashSetError::ElementAlreadyExists)
+        ));
+
+        let element_4_93 = 93.to_biguint().unwrap();
+        let element_4_65 = 64.to_biguint().unwrap();
+        let element_4_72 = 72.to_biguint().unwrap();
+        let element_4_15 = 15.to_biguint().unwrap();
+        hs.insert(&element_4_93, 3).unwrap();
+        hs.insert(&element_4_65, 3).unwrap();
+        hs.insert(&element_4_72, 3).unwrap();
+        hs.insert(&element_4_15, 3).unwrap();
+        assert_eq!(hs.contains(&element_4_93, 4).unwrap(), true);
```

Comment:
shouldn't this return false since sequence number of 3 is expired?

Reply:
At this point, `element_4_93` wasn't marked with any sequence number. It was just inserted. So its sequence number is None. The value `3` provided in the insert os not assigned - it's just used for potentially vacating the cell (which doesn't happen in this test case though, but I just wanted to replicate the mechanism we have in the program).